pkg/controller/deployment_rolling.go

```diff
@@ -44,6 +44,13 @@ var (
 
 // rolloutRolling implements the logic for rolling a new machine set.
 func (dc *controller) rolloutRolling(d *v1alpha1.MachineDeployment, isList []*v1alpha1.MachineSet, machineMap map[types.UID]*v1alpha1.MachineList) error {
+
+	clusterAutoscalerScaleDownAnnotations := make(map[string]string)
+	clusterAutoscalerScaleDownAnnotations[v1alpha1.ClusterAutoscalerScaleDownDisabledAnnotationKey] = v1alpha1.ClusterAutoscalerScaleDownDisabledAnnotationValue
+
+	// We do this avoid accidentally deleting the user provided annotations.
+	clusterAutoscalerScaleDownAnnotations[v1alpha1.ClusterAutoscalerScaleDownDisabledAnnotationByMCMKey] = v1alpha1.ClusterAutoscalerScaleDownDisabledAnnotationByMCMValue
```

Comment:
Hi Amshu, I believe, if we add the `by-mcm` annotation on the MachineDeployment's NodeTemplate, the worker-extension will overwrite it back to original.

Reply:
Thanks, understood. Then a modified follow up question. How about using the `nodeTemplate` mechanism in the `machinesets`? I.e. adding the annotations in to the `nodeTemplate` in the `machinesets`?

The reason for asking this is that, if we annotate the nodes directly, then only existing nodes get annotated. If in the meantime any of the `machinesets` get scaled up (probably, manually), the new nodes won't get the annotation unless a `machinedeploy` reconciliation is executed. Admittedly, that is a contrived scenario and not at all a common or important one. But thought it might be nice if we can avoid that edge case.

Reply:
>Then a modified follow up question. How about using the nodeTemplate mechanism in the machinesets? I.e. adding the annotations in to the nodeTemplate in the machinesets?

In this case, it'd be overwritten again. If there are any annotations on the `nodeTemplate` of the MachineSet but not on the `nodeTemplate` of the MachineDeployment, the deployment-controller shall overwrite it back to original. 

At the moment, every join of the node re-enqueues the Machine -> MachiineSet -> MachineDeployment, hence new nodes should immediately be annotated they join.  

Also, to mention, cluster-autoscaler prefers slow-scaledown, it has certain timeouts like `scale-down-delay-after-add` before which it decides to trigger scale-down of a given node. This should also complement the logic.

Reply:
> In this case, it'd be overwritten again. If there are any annotations on the `nodeTemplate` of the MachineSet but not on the `nodeTemplate` of the MachineDeployment, the deployment-controller shall overwrite it back to original.

Yes. But this argument is not as strong as the argument against updating the `nodeTemplate` on the `MachineDeployment` because modifying the worker-extension is not desirable. We could modify the existing code code path in the deployment controller to make it not overwrite the annotations rather than introduce a new place and mechanism for of annotating the nodes in the code. Similar logic already exists for not overwriting manually maintained annotations on the node. WDYT?

Reply:
> However, we could modify the existing code code path in the deployment controller to make it not overwrite the annotations rather than introduce a new place and mechanism for annotating the nodes in the code.

There is already a mechanism to add the `PreferNoSchedule` taints directly on the node-objects once roll-out starts, this logic takes the same path. 

> Similar logic already exists for not overwriting manually maintained annotations on the node. WDYT?

MCM doesn't specifically ignore the manually added annotations on the node, it only maintains the ones which are part of the `nodeTemplate`, so there is actually no mechanism of ignoring the annotations in `nodeTemplate`.

Also, if MCM starts ignoring only `cluster-autoscaler.kubernetes.io/scale-down-disabled: "True"` in the machine-set's `nodeTemplate`, then the same annotation put by the end-user on the machine-deployment's `nodeTemplate` will also be ignored.  Then we need to find a way to ignore the tuple of annotations in `nodeTemplate`, which looks a little complex to me, and output would probably be not worth changing the entire approach.

Reply:
>  if MCM starts ignoring only cluster-autoscaler.kubernetes.io/scale-down-disabled: "True" in the machine-set's nodeTemplate, then the same annotation put by the end-user on the machine-deployment's nodeTemplate will also be ignored.

I was not suggesting to not use the `by-mcm` annotation. That approach would work equally well if the nodes are annotated directly or via the `nodeTemplate`.

> There is already a mechanism to add the `PreferNoSchedule` taints directly on the node-objects once roll-out starts, this logic takes the same path.

I would suggest to change that also to use the `nodeTemplate` path to streamline the interaction with nodes. But not in this PR, of course.

But anyway, my point about `nodeTemplate` was a nice-to-have to avoid too many paths for modifying the node objects and that it makes life difficult to simulate all the scenarios to be tested.